controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -115,7 +115,7 @@ public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher
                 this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
                 return this.client;
             } else {
-                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
+                log.info("Former ZK connection string: {}. New ZK connection string {}.", this.connectString, connectString);
```

Comment:
We are transforming a precondition into a log message. We are also keeping the precondition checks for all other parameters. If we are to ignore changes to the input, then we need to remove all these precondition equality checks.

One alternative that looks cleaner to me is something like the following:

```
        try {
                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
                Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
                Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
                Preconditions.checkNotNull(watcher, "watcher");
                this.client.register(watcher);
         } catch (IllegalArgumentException e) {
                log.warn("Input argument for new ZooKeeper client changed with respect to existing client", e);
                this.client.close();
         } finally {
                return this.client;
         }
```

In the case something doesn't match, we make sure that the session is invalid from this point onwards and create no confusion.